src/System.Text.Primitives/System/Text/Encoders/Utf8Utility.cs

```diff
@@ -0,0 +1,357 @@
+                        }
+                        else
+                        {
+                            byte fourthByte = data[3];
+                            if (IsUtf8ContinuationByte(fourthByte))
+                            {
+                                // Well-formed four-byte sequence.
+                                scalar |= (((uint)thirdByte & 0x3FU) << 6) | ((uint)fourthByte & 0x3FU);
+                                scalarValue = UnicodeScalar.CreateWithoutValidation(scalar);
+                                numBytesConsumed = 4;
+                                return SequenceValidity.WellFormed;
+                            }
+                            else
+                            {
+                                // Three bytes of an improperly terminated multi-byte sequence.
+                                goto InvalidThreeByteSequence;
+                            }
+                        }
+                    }
+                    else
+                    {
+                        // Two bytes of improperly terminated multi-byte sequence.
+                        goto InvalidTwoByteSequence;
+                    }
+                }
+            }
+
+            // Everything below here is error handling.
+
+            InvalidOneByteSequence:
+            numBytesConsumed = 1;
+            return SequenceValidity.Invalid;
+
+            InvalidTwoByteSequence:
+            OverlongOutOfRangeOrSurrogateSequence:
+            numBytesConsumed = 2;
+            return SequenceValidity.Invalid;
+
+            InvalidThreeByteSequence:
+            numBytesConsumed = 3;
+            return SequenceValidity.Invalid;
+
+            OneByteOfIncompleteMultiByteSequence:
+            numBytesConsumed = 1;
+            return SequenceValidity.Incomplete;
+
+            TwoBytesOfIncompleteMultiByteSequence:
+            numBytesConsumed = 2;
+            return SequenceValidity.Incomplete;
+
+            ThreeBytesOfIncompleteMultiByteSequence:
+            numBytesConsumed = 3;
+            return SequenceValidity.Incomplete;
+        }
+
+        /// <summary>
+        /// Calculates the number of UTF-16 code units (<see cref="char"/>s) that would result from converting
+        /// the provided UTF-8 string to UTF-16 representation.
+        /// </summary>
+        /// <param name="inputBuffer">The buffer containing UTF-8 text.</param>
+        /// <param name="charCount">
+        /// If this method returns <see langword="true"/>, contains the equivalent <see cref="char"/> count of the input string.
+        /// If this method returns <see langword="false"/>, the value is undefined.
+        /// </param>
+        /// <returns>
+        /// <see langword="true"/> on success, <see langword="false"/> if the input is not a well-formed UTF-8 string.
+        /// </returns>
+        public static bool TryGetUtf16CharCount(ReadOnlySpan<byte> inputBuffer, out int charCount)
```

Comment:
It's strange to see this on Utf8Util type. Also, inputBuffer should be called utf8?